### ISSUE TITLE
The effeciency of GATE

### ISSUE THREAD

**Issue by LuoXueling** [manujosephv/pytorch_tabular#183]
I am using the new SOTA model GATE in the 1.0.1 release. As described in the [paper](https://arxiv.org/abs/2207.08548), GATE has fewer FLOPs compared to NODE (and some other models)  when the numbers of parameters are similar.

However, when I take all `pytorch_tabular` models as my baselines, I find that GATE is extremely slow both on CPU (i9-11900K) and GPU (RTX 3090 + CUDA 11.6) compared to other models. 

Here is my experiment on GPU:

```python
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression
from pytorch_tabular import TabularModel
from pytorch_tabular.models import NodeConfig, GatedAdditiveTreeEnsembleConfig
from pytorch_tabular.config import DataConfig, OptimizerConfig, TrainerConfig

cont_cols = [f"cont_{i}" for i in range(100)]
data = make_regression(n_samples=1024, n_features=100, n_targets=1, random_state=0)
df = pd.DataFrame(
    data=np.hstack((data[0], data[1].reshape(-1, 1))), columns=cont_cols + ["target"]
)
data_config = DataConfig(
    target=["target"],
    continuous_cols=cont_cols,
    num_workers=4,
)
trainer_config = TrainerConfig(
    accelerator="auto",
)
optimizer_config = OptimizerConfig()

node_config = NodeConfig(task="regression", embed_categorical=True)
node_model = TabularModel(
    data_config=data_config,
    model_config=node_config,
    optimizer_config=optimizer_config,
    trainer_config=trainer_config,
)
gate_config = GatedAdditiveTreeEnsembleConfig(task="regression", chain_trees=False)
gate_model = TabularModel(
    data_config=data_config,
    model_config=gate_config,
    optimizer_config=optimizer_config,
    trainer_config=trainer_config,
)

node_model.fit(train=df, max_epochs=5)
gate_model.fit(train=df, max_epochs=5)
```

And the following is the output:

```shell
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_tabular/models/mixture_density/mdn.py:25: UserWarning: Wandb not installed. WandbLogger will not work.
  warnings.warn("Wandb not installed. WandbLogger will not work.")
2023-04-02 13:19:02,063 - {pytorch_tabular.tabular_model:102} - INFO - Experiment Tracking is turned off
2023-04-02 13:19:02,113 - {pytorch_tabular.tabular_model:102} - INFO - Experiment Tracking is turned off
Global seed set to 42
2023-04-02 13:19:02,129 - {pytorch_tabular.tabular_model:465} - INFO - Preparing the DataLoaders
2023-04-02 13:19:02,129 - {pytorch_tabular.tabular_datamodule:286} - INFO - Setting up the datamodule for regression task
2023-04-02 13:19:02,141 - {pytorch_tabular.tabular_model:508} - INFO - Preparing the Model: NODEModel
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_tabular/models/node/odst.py:102: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  bin_codes = (indices.view(1, -1) // offsets.view(-1, 1) % 2).to(torch.float32)
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_tabular/models/node/node_model.py:116: UserWarning: Ignoring head config because NODE has a specific head which subsets the tree outputs
  warnings.warn("Ignoring head config because NODE has a specific head which subsets the tree outputs")
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_tabular/models/base_model.py:126: UserWarning: Wandb is not installed. Please install wandb to log logits. You can install wandb using pip install wandb or install PyTorch Tabular using pip install pytorch-tabular[all]
  warnings.warn(
2023-04-02 13:19:02,167 - {pytorch_tabular.models.node.node_model:82} - INFO - Data Aware Initialization of NODE using a forward pass with 2000 batch size....
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_tabular/models/node/odst.py:143: UserWarning: Data-aware initialization is performed on less than 1000 data points. This may cause instability.To avoid potential problems, run this model on a data batch with at least 1000 data samples.You can do so manually before training. Use with torch.no_grad() for memory efficiency.
  warn(
2023-04-02 13:19:04,613 - {pytorch_tabular.tabular_model:264} - INFO - Preparing the Trainer
Auto select gpus: [0]
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
2023-04-02 13:19:04,647 - {pytorch_tabular.tabular_model:566} - INFO - Training Started
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:604: UserWarning: Checkpoint directory /home/xlluo/hdd/ML-fracture/test/gate/saved_models exists and is not empty.
  rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
┏━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃   ┃ Name             ┃ Type             ┃ Params ┃
┡━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│ 0 │ _backbone        │ NODEBackbone     │  1.8 M │
│ 1 │ _embedding_layer │ Embedding1dLayer │    200 │
│ 2 │ _head            │ Lambda           │      0 │
│ 3 │ loss             │ MSELoss          │      0 │
└───┴──────────────────┴──────────────────┴────────┘
Trainable params: 1.8 M                                                         
Non-trainable params: 769                                                       
Total params: 1.8 M                                                             
Total estimated model params size (MB): 7                                       
`Trainer.fit` stopped: `max_epochs=5` reached.
Epoch 4/4  ━━━━━━━━━━━━━━━━━ 17/17 0:00:00 • 0:00:00 71.84it/s loss: 2.09e+04   
                                                               train_loss:      
                                                               15978.538        
                                                               valid_loss:      
                                                               22757.045        
                                                               valid_mean_squar…
                                                               22757.045        
                                                               train_mean_squar…
                                                               20967.875        
2023-04-02 13:19:08,261 - {pytorch_tabular.tabular_model:568} - INFO - Training the model completed
2023-04-02 13:19:08,261 - {pytorch_tabular.tabular_model:1207} - INFO - Loading the best model
Global seed set to 42
2023-04-02 13:19:08,275 - {pytorch_tabular.tabular_model:465} - INFO - Preparing the DataLoaders
2023-04-02 13:19:08,275 - {pytorch_tabular.tabular_datamodule:286} - INFO - Setting up the datamodule for regression task
2023-04-02 13:19:08,287 - {pytorch_tabular.tabular_model:508} - INFO - Preparing the Model: GatedAdditiveTreeEnsembleModel
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_tabular/models/base_model.py:126: UserWarning: Wandb is not installed. Please install wandb to log logits. You can install wandb using pip install wandb or install PyTorch Tabular using pip install pytorch-tabular[all]
  warnings.warn(
2023-04-02 13:19:08,478 - {pytorch_tabular.models.gate.gate_model:221} - INFO - Data Aware Initialization of T0
2023-04-02 13:19:08,603 - {pytorch_tabular.tabular_model:264} - INFO - Preparing the Trainer
Auto select gpus: [0]
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
2023-04-02 13:19:08,617 - {pytorch_tabular.tabular_model:566} - INFO - Training Started
/home/xlluo/anaconda3/envs/mlfatigue/lib/python3.8/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:604: UserWarning: Checkpoint directory /home/xlluo/hdd/ML-fracture/test/gate/saved_models exists and is not empty.
  rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
┏━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃   ┃ Name             ┃ Type                       ┃ Params ┃
┡━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│ 0 │ _backbone        │ GatedAdditiveTreesBackbone │  706 K │
│ 1 │ _embedding_layer │ Embedding1dLayer           │    200 │
│ 2 │ _head            │ CustomHead                 │     54 │
│ 3 │ loss             │ MSELoss                    │      0 │
└───┴──────────────────┴────────────────────────────┴────────┘
Trainable params: 706 K                                                         
Non-trainable params: 0                                                         
Total params: 706 K                                                             
Total estimated model params size (MB): 2                                       
`Trainer.fit` stopped: `max_epochs=5` reached.
Epoch 4/4  ━━━━━━━━━━━━━━━━━ 17/17 0:00:06 • 0:00:00 2.63it/s loss: 1.97e+04    
                                                              train_loss:       
                                                              25493.67          
                                                              valid_loss:       
                                                              21327.895         
                                                              valid_mean_square…
                                                              21327.895         
                                                              train_mean_square…
                                                              19608.16          
2023-04-02 13:19:44,075 - {pytorch_tabular.tabular_model:568} - INFO - Training the model completed
2023-04-02 13:19:44,075 - {pytorch_tabular.tabular_model:1207} - INFO - Loading the best model
```

We could see from the progress bar that GATE is much slower than NODE with default parameters, even when NODE has a higher number of parameters. In my practice, it takes hours to train GATE for 100 epochs, which is really a bottleneck when evaluating my baseline models.

For your reference, I am using `torch==1.13.0+cu116`, `pytorch_tabular==1.0.1`, python 3.8, and Ubuntu 18.04.

**Comment by manujosephv** [comment]
Although GATE does have lower FLOPS, I hypothesise that the runtime is mainly because of a `sort` operation in `entmax` activation. Can you try using another activation function like sparsemax and sparsemoid ? Right now I don't have access to my computer and unable to tell you exact parameter names. But in the documentation, there will be two parameters which control binning activation and feature masking activation. Default values are entmoid and entmax. Can you try changing them to sparsemoid and sparsemax?

**Comment by manujosephv** [comment]
Another thing is that the default parameters in GATE are slightly more heavier, and are not entirely necessary for all datasets. You can get good models by using smaller models as well. Key parameters to tweak would be tree depth and number of trees, both of which have a significant impact on runtimes.

**Comment by LuoXueling** [comment]
Thanks for your kind response! I run three experiments as follows (on the CPU this time because my GPU is busy). To be concise, parameters and activation functions do impact runtimes. I would consider decreasing these parameters. If it is not a bug or implementation issue, you could close this issue.

1. With default parameters as described in the paper
```python
gate_config = GatedAdditiveTreeEnsembleConfig(task="regression")
```
It has `Trainable params: 1.6 M `, and the progress bar shows `Epoch 4/4  ━━━━━━━━━━━━━━━━━ 17/17 0:00:11 • 0:00:00 1.56it/s`.

2. With "Lite" parameters in the paper
```python
gate_config = GatedAdditiveTreeEnsembleConfig(
    task="regression",
    chain_trees=False,
    tree_depth=5,
    num_trees=30,
    gflu_stages=4,
)
```
It has `Trainable params: 757 K`, and the progress bar shows `Epoch 4/4  ━━━━━━━━━━━━━━━━━ 17/17 0:00:10 • 0:00:00 1.74it/s`.

3. As suggested, changing `binning_activation` to `sparsemoid` and `feature_mask_function` to `sparsemax`
```python
gate_config = GatedAdditiveTreeEnsembleConfig(
    task="regression",
    chain_trees=False,
    tree_depth=5,
    num_trees=30,
    gflu_stages=4,
    binning_activation="sparsemoid",
    feature_mask_function="sparsemax",
)
```
It has `Trainable params: 757 K`, and the progress bar shows `Epoch 4/4  ━━━━━━━━━━━━━━━━━ 17/17 0:00:08 • 0:00:00 2.13it/s`.

I could further change `tree_depth=4` and `num_trees=20`. This time it has 400 K params and the progress bar shows `Epoch 4/4  ━━━━━━━━━━━━━━━━━ 17/17 0:00:03 • 0:00:00 5.46it/s`

**Comment by manujosephv** [comment]
#196 Changes the default to sparsemax. Closing the issue
